Adding native ads support to AppLovin adapter.

santoshbagadi opened the pull request.

santoshbagadi committed:
Adding native ads support to AppLovin adapter.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,107 @@
+public class AppLovinNativeAdListener
+        implements AppLovinNativeAdLoadListener, AppLovinNativeAdPrecacheListener {
+
+    private AppLovinNativeAdapter adapter;
```
Lets follow Android/oldschool Java/ (and now AppLovin) conventions of prefixing fields with `mXXX`. So this would be `mAdapter`. I don't love it either, but I see that's how Android design theri classes.

thomasmso replied:
You can also mark all of these as `final`, since they are assigned in the constructor.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,107 @@
+                                    MediationNativeListener nativeListener,
+                                    AppLovinSdk sdk,
+                                    Context context) {
+        this.adapter = adapter;
```
Then you won't have to use `this.xxx`, if using `m` prefixes.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,107 @@
+import java.lang.ref.WeakReference;
+import java.util.List;
+
+public class AppLovinNativeAdListener
```
You don't need to make this public for the world right? Since it' sonly access within this package, you can make the scope smaller by removing (public) to default into package-level access.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdapter.java:
```diff
@@ -0,0 +1,54 @@
+import com.google.android.gms.ads.mediation.MediationNativeListener;
+import com.google.android.gms.ads.mediation.NativeMediationAdRequest;
+
+public class AppLovinNativeAdapter
```
Not sure if it needs to be `public`.

santoshbagadi replied:
This needs to be `public` so that the AdMob SDK will be able to instanciate this class using reflection.

thomasmso replied:
I'm pretty sure even if we removed `public`, Dalvik / JVM can still access the class even from a completely different package, it's the methods/fields that will be more annoying as they are scope-protected. BUT, since we mark adapters public everywhere, we should keep `public` :). Convention > Convention

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,99 @@
+
+import java.util.ArrayList;
+
+public class AppLovinNativeAdMapper extends NativeAppInstallAdMapper {
```
Not sure if it needs to be `public`.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdapter.java:
```diff
@@ -0,0 +1,54 @@
+
+public class AppLovinNativeAdapter
+        implements MediationNativeAdapter {
+    protected static final String TAG = AppLovinNativeAdapter.class.getSimpleName();
```
You don't need `protected`, it will have package-level access still.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdapter.java:
```diff
@@ -0,0 +1,54 @@
+    protected static final String TAG = AppLovinNativeAdapter.class.getSimpleName();
+
+    public static final String KEY_EXTRA_AD_ID = "ad_id";
+    public static final String KEY_EXTRA_CAPTION_TEXT = "caption_text";
```
Why are these keys in this class, when they are only used in another file?

santoshbagadi replied:
These keys will be used by the publisher to get the extras from the extras bundle.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdapter.java:
```diff
@@ -0,0 +1,54 @@
+                && !nativeMediationAdRequest.isAppInstallAdRequested()) {
+            Log.w(TAG, "Failed to request native ad. App install format needs to be requested");
+            mediationNativeListener
+                    .onAdFailedToLoad(AppLovinNativeAdapter.this, AdRequest.ERROR_CODE_INVALID_REQUEST);
```
Why do you need `AppLovinNativeAdapter.this` when `this` could do?

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdapter.java:
```diff
@@ -0,0 +1,54 @@
+            return;
+        }
+
+        final AppLovinSdk sdk = AppLovinSdk.getInstance(context);
```
I think we have a Utility method, to retrieve the AppLovinSdk. b/c the SDK key could either be passed in via `serverParameters`, or exist in the publisher's AndroidManifest. Check other adapters for example.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdapter.java:
```diff
@@ -0,0 +1,54 @@
+
+        final AppLovinSdk sdk = AppLovinSdk.getInstance(context);
+        AppLovinNativeAdListener listener = new AppLovinNativeAdListener(
+                AppLovinNativeAdapter.this,
```
Why not just `this`?

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,107 @@
+        this.adapter = adapter;
+        this.nativeListener = nativeListener;
+        this.appLovinSdk = sdk;
+        contextWeakReference = new WeakReference<>(context);
```
👍

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,107 @@
+    public void onNativeAdsLoaded(List<AppLovinNativeAd> nativeAds) {
+        if (nativeAds.size() > 0 && isValidNativeAd(nativeAds.get(0))) {
+            appLovinSdk.getNativeAdService()
+                    .precacheResources(nativeAds.get(0), AppLovinNativeAdListener.this);
```
Why not just `this`?

thomasmso replied:
We'll never actually return an empty array of native ads, but I like the paranoia! Feel free to keep it in.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,107 @@
+    private static boolean isValidNativeAd(AppLovinNativeAd nativeAd) {
+        return nativeAd.isImagePrecached() && nativeAd.getImageUrl() != null
+                && nativeAd.getIconUrl() != null && nativeAd.getTitle() != null
+                && nativeAd.getDescriptionText() != null && nativeAd.getCtaText() != null;
```
In regards to `isImagePrecached()` - It's not always true that AppLovin SDK precaches the image. Check out setting `NATIVE_ADS_IN_BATCH_TO_PRECACHE`. in `SettingTypes.java`.

santoshbagadi replied:
I'm explicitly requesting the ads to be pre-cached once and ad is loaded. To create a media view using an image, we need to have the images pre-cached.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,107 @@
+            }
+        });
+
+    }
```
"There's an extra new line"

- OCD guy

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,99 @@
+
+        ArrayList<NativeAd.Image> images = new ArrayList<>(1);
+        AppLovinNativeAdImage image;
+        AppLovinNativeAdImage icon;
```
Why declare them, then assign later. We generally declare vars when needed.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,99 @@
+        Bundle extraAssets = new Bundle();
+        extraAssets.putLong(AppLovinNativeAdapter.KEY_EXTRA_AD_ID, nativeAd.getAdId());
+        extraAssets.putString(
+                AppLovinNativeAdapter.KEY_EXTRA_CAPTION_TEXT, nativeAd.getCaptionText());
```
`Bundle` employs a builder pattern so you can technically do:

```
Bundle extraAssets = new Bundle().putLong(AppLovinNativeAdapter.KEY_EXTRA_AD_ID, nativeAd.getAdId())
     .putString(...)
    .putThat(...);
```

santoshbagadi replied:
I don't think that is true. [`putLong(key, value)`](https://developer.android.com/reference/android/os/BaseBundle.html#putLong(java.lang.String,%20long)) (or any other `Bundle.put...(key, value)`) is not returning a bundle, so we won't be able to chain multiple puts together.

thomasmso replied:
Ah I must've been thinking about something else..

thomasmso replied:
Intents!

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,99 @@
+                AppLovinNativeAdapter.KEY_EXTRA_CAPTION_TEXT, nativeAd.getCaptionText());
+
+        setOverrideClickHandling(false);
+        setOverrideImpressionRecording(false);
```
Question: What do these calls do?

santoshbagadi replied:
These calls tell AdMob SDK whether or not our SDK will be handling clicks and impressions. Since we are setting then to `false` here, AdMob calls the `recordImpression()` and `handleClick(View view)` methods below.

santoshbagadi committed:
Adding style changes.

santoshbagadi commented:
Changes made.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,109 @@
+
+    private final AppLovinNativeAdapter mAdapter;
+    private final MediationNativeListener mNativeListener;
+    private final AppLovinSdk mAppLovinSdk;
```
Pretty OCD, but in our other adapters, we call this field `mSdk`, could you perhaps just keep it consistent with them?

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,109 @@
+        // Create a native ad.
+        Context context = mContextWeakReference.get();
+        if (context == null) {
+            Log.w(AppLovinNativeAdapter.TAG, "Failed to create mapper. Context is null.");
```
Should this be a warning or error?

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,109 @@
+        if (nativeAds.size() > 0 && isValidNativeAd(nativeAds.get(0))) {
+            mAppLovinSdk.getNativeAdService().precacheResources(nativeAds.get(0), this);
+        } else {
+            Log.w(AppLovinNativeAdapter.TAG,
```
Lets mark this as error, since theoretically our SDK will never return empty list of native ads, if it 204s, it'll just call back the `onNativeAdsFailedToLoad()` callback.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,109 @@
+
+    @Override
+    public void onNativeAdVideoPrecachingFailed(AppLovinNativeAd ad, final int errorCode) {
+        notifyAdFailure(AppLovinUtils.toAdMobErrorCode(errorCode));
```
Can the mapper take in the raw `Uri` and stream that? I'd probably log a error here.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java:
```diff
@@ -0,0 +1,109 @@
+
+    @Override
+    public void onNativeAdImagePrecachingFailed(AppLovinNativeAd ad, final int errorCode) {
+        notifyAdFailure(AppLovinUtils.toAdMobErrorCode(errorCode));
```
Can we log an error here? Basically we log all the callbacks, it's good for debugging and consistency (if you check out our other adapters).

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,96 @@
+        mediaView.setLayoutParams(layoutParams);
+
+        ArrayList<NativeAd.Image> images = new ArrayList<>(1);
+        Drawable imageDrawable = Drawable.createFromPath(Uri.parse(nativeAd.getImageUrl()).getPath());
```
There's a lot nested and chained calls here, could you improve readability by extracting some of it? LIke `Uri.parse(nativeAd.getImageUrl()).getPath()`

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,96 @@
+        setMediaView(mediaView);
+        setStarRating(nativeAd.getStarRating());
+
+        Bundle extraAssets = new Bundle();
```
The point of `extraAssets`? Is it being used anywhere?

santoshbagadi replied:
Any assets that cannot be directly mapped to an AdMob ad can be passed to the publisher using the extras bundle.

The publisher would use the keys in `AppLovinNativeAdapter` class to access the assets from the bundle. These keys need to be documented (See [Facebook adapter documentation](https://developers.google.com/admob/android/mediation/facebook#using_native_ads) for example)

thomasmso replied:
1. I still don't get how the publisher can access this `Bundle`. It looks like you're just creating a `Bundle`, adding the ad id and caption into it, and it'll get deallocated, right? Are you missing a superclass call like a `setBundleExtras(xxx)` or something?
2. How come you're not passing in the video? Are we going to add support for videos in this release, or in a future one?

santoshbagadi replied:
Ah, I missed the `setExtras(Bundle)`call. Added it now. The publisher would be able to now get them using the [`getExtras()`](https://developers.google.com/android/reference/com/google/android/gms/ads/formats/NativeAppInstallAd.html#getExtras()) method on `NativeAppInstallAd`.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,96 @@
+    @Override
+    public void recordImpression() {
+        super.recordImpression();
+        mNativeAd.trackImpression();
```
Do you know if AdMob ensures this is called **once and only once**? If you are not sure, could we surround with an `AtomicBoolean`?

santoshbagadi replied:
Yes, this should only get called once.

thomasmso replied:
Could we remove the unnecessary call to super? Since we've removed it on iOS

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,96 @@
+    /**
+     * A {@link NativeAd.Image} class used to map AppLovin native image to AdMob native image.
+     */
+    private class AppLovinNativeAdImage extends NativeAd.Image {
```
Since this does not require any of the enclosing class's instance fields, you can probably mark it as static too, so each instance of `AppLovinNativeAdMapper` doesn't need a separate nested class of this.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,96 @@
+    private class AppLovinNativeAdImage extends NativeAd.Image {
+
+        private Drawable mDrawable;
+        private Uri mUri;
```
Can you mark these guys as final?

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdMapper.java:
```diff
@@ -0,0 +1,96 @@
+
+        AppLovinNativeAdImage(Uri uri, Drawable drawable) {
+            this.mDrawable = drawable;
+            this.mUri = uri;
```
You don't need `this` since we're using the `m` prefixing convention.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdapter.java:
```diff
@@ -0,0 +1,49 @@
+                                final Bundle mediationExtras) {
+        if (nativeMediationAdRequest.isContentAdRequested()
+                && !nativeMediationAdRequest.isAppInstallAdRequested()) {
+            Log.w(TAG, "Failed to request native ad. App install format needs to be requested");
```
Could we log this as an error?

santoshbagadi committed:
Logging errors and style changes.

santoshbagadi committed:
More style changes

santoshbagadi reviewed (commented):
Changes made.

thomasmso commented on ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdapter.java:
```diff
@@ -0,0 +1,49 @@
+                                final Bundle serverParameters,
+                                final NativeMediationAdRequest nativeMediationAdRequest,
+                                final Bundle mediationExtras) {
+        if (nativeMediationAdRequest.isContentAdRequested()
```
Why don't we allow content ads?

thomasmso replied:
But on iOS we do?

santoshbagadi replied:
We don't support content ads on either platform. On iOS we are supporting Unified and App Install ad formats. 

Unified is a new format added by AdMob which is not available on Android yet.

santoshbagadi committed:
Setting extras to the native ad.

santoshbagadi commented:
Changes made.

santoshbagadi committed:
Removing supre call and isPrecached check.

- Since we are requesting for the ads to be explicity precached, no need to check for that.

santoshbagadi commented:
Changes made.
